### PR TITLE
PR: Fix cog menu not showing in the variable explorer when more than one IPython console is opened

### LINKED ISSUE #7704: No cog menu in variable explorer widget toolbar when more than one Ipython console is opened
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [x] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [x] Reset preferences with ``spyder --reset``
    * [x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

The cog menu appears only for the variable explorer associated with the last created IPython console.

### What steps reproduce the problem?

1. Create a new IPython console so that there is at least 2 consoles opened
2. The variable explorer associated with the last IPython console has a cog menu
3. The variable explorer associated with the previous IPython consoles does not have a cog menu

cog_menu_not_duplicated

### What is the expected output? What do you see instead?

The cog menu should remain visible on the variable explorer toolbar for all the IPython consoles. 

## Versions

* Spyder version: Spyder 4.0.0.dev0
* Python version: 3.6.5 64 bits
* Qt version: 5.9.3
* PyQt version: 5.9.2
* Operating System name/version: Windows 10

### Dependencies

```
pyflakes >=0.6.0  :  1.6.0 (OK)
pycodestyle >=2.3 :  2.3.1 (OK)
pygments >=2.0    :  2.2.0 (OK)
sphinx >=0.6.6    :  1.7.1 (OK)
rope >=0.9.4      :  0.10.7 (OK)
jedi >=0.11.0     :  0.11.1 (OK)
nbconvert >=4.0   :  5.3.1 (OK)
pandas >=0.13.1   :  0.22.0 (OK)
numpy >=1.7       :  1.13.0 (OK)
sympy >=0.7.3     :  1.1.1 (OK)
cython >=0.21     :  0.27.3 (OK)
qtconsole >=4.2.0 :  4.4.0.dev (OK)
IPython >=4.0     :  6.2.1 (OK)
matplotlib >=2.0.0:  2.2.0 (OK)
pylint >=0.25     :  1.8.3 (OK)
```

### PR DESCRIPTION
**Pull request by jnsebgosselin**
### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] Wrote at least one-line docstrings for any new functions
* [ ] Added at least one unit test covering the changes, if at all possible
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] Included a screenshot, if this PR makes any visible changes to the UI

## Description of Changes

The layout of the `VariableExplorer` plugin consists mainly of a `StackWidget`.  When an IPython console is created, a corresponding `NamespaceBrowser` is instantiated and added to the stack.

The cog menu button (cmb) is created by the variable explorer plugin and passed as a class argument when instantiating a new namespace browser (nsb ), which is later added to its layout during the setup. So this means that the cmb is shared among all the nsb of the stack.

The problem is that, when the current IPython console change, and so the current nsb of the stack widget, there is no mechanism to update the cmb  actions and to move it from the previous current nsb layout to that of the new current nsb. In order to solve this issue, this PR:

- Add the code infrastructure to update the cmb actions when the current nsb of the stack change.
- Add the code infrastructure to move the cmb to the current nsb layout.

The changes at the plugin layer are minimal and use the public interface already in place for plugins. Almost all changes were made at the widget layer.

### Issue(s) Resolved

Fixes #7704

### CODE REVIEW

**Comment by jnsebgosselin** [comment]
@ccordoba12 If you want a test to cover this, could you please tell me where this test should go?

File: spyder/plugins/variableexplorer.py [hunk 1]
```diff
@@ -164,9 +168,7 @@ def add_shellwidget(self, shellwidget):
         shellwidget_id = id(shellwidget)
         if shellwidget_id not in self.shellwidgets:
             self.options_button.setVisible(True)
-            nsb = NamespaceBrowser(self,
-                                   options_button=self.options_button,
-                                   plugin_actions=[self.undock_action])
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer.py]: Why did you remove the `undock_action` from here?

**Comment by ccordoba12** [comment]
> If you want a test to cover this

Don't worry, I don't think a test is necessary (or even doable) for this.

File: spyder/plugins/variableexplorer.py (hunk 1 above)
Comment (jnsebgosselin) [reply spyder/plugins/variableexplorer.py]: Presently, actions are added to the menu of the `options_button` during the setup of the `NamespaceBrowser` widget. This is a problem because the `options_button` is shared among all the namespace browsers of the variable explorer plugin (which are added to a `StackWidget`). Therefore, the `options_button` ends up with the actions of the last namespace browser added to the plugin and the actions are not refreshed when the current namespace browser changes.

To fix this issue, I'm proposing to make the  variable explorer plugin responsible to add the actions to the `options_button` instead of doing it in the widget . Therefore, the `plugin_actions` argument is kept here only as a convenience when using the  `NamespaceBrowser` widget outside of the `VariableExplorer` plugin. As you can see, the `NamespaceBrowser.plugin_actions attribute` is only use if the `options_button` argument is `None` when instantiating the `NamespaceBrowser` widget. Please see https://github.com/spyder-ide/spyder/pull/7710/files#diff-6b96bad69031242aedccaa716731ff49R217

So, since we do not override the actions in the widget anymore, we do not need to pass the `undock_action` action when instantiating the `NamespaceBrowser` in the `VariableExplorer` plugin, because this is already added by default to the menu of the `options_button` when the plugin is initialize. See https://github.com/spyder-ide/spyder/blob/master/spyder/api/plugins.py#L102.

Besides, each time the current namespace browser change, the actions of the `options_buttons` are refreshed by the variable explorer plugin using the [refresh_actions](https://github.com/spyder-ide/spyder/blob/2208ff88834bc575383c61d2bf8096fb52ad5642/spyder/api/plugins.py#L209) method of the plugin api, which automatically append the `undock_action` to the plugin actions.

Sorry for the long answer, I hope my explanations were ok.

Comment (ccordoba12) [reply spyder/plugins/variableexplorer.py]: Ok, thanks for the really nice explanation!
